lighttpd: use autoconf instead of patching configure

fbrosson opened the pull request.

waddlesplash commented on www-servers/lighttpd/lighttpd-1.4.39.recipe:
```diff
@@ -14,7 +14,7 @@ SOURCE_URI="https://download.lighttpd.net/lighttpd/releases-1.4.x/lighttpd-$port
 PATCHES="lighttpd-$portVersion.patch"
 
-ARCHITECTURES="x86_gcc2 ?x86 ?x86_64 ?arm ?ppc"
+ARCHITECTURES="x86_gcc2 !x86 !x86_64 ?arm ?ppc"
```
Wait, so it actually is broken on these architectures?

fbrosson replied:
Yep, it is currently broken on x86 and x86_64. Compilation seems ok, but the link fails:
```
connections.o: In function `connection_accept':
/sources/lighttpd-1.4.39/src/connections.c:1132: undefined reference to `accept'
connections.o: In function `connection_state_machine':
/sources/lighttpd-1.4.39/src/connections.c:1585: undefined reference to `shutdown'
/sources/lighttpd-1.4.39/src/connections.c:1415: undefined reference to `shutdown'
network.o: In function `network_server_init':
/sources/lighttpd-1.4.39/src/network.c:271: undefined reference to `setsockopt'
/sources/lighttpd-1.4.39/src/network.c:258: undefined reference to `socket'
/sources/lighttpd-1.4.39/src/network.c:371: undefined reference to `connect'
/sources/lighttpd-1.4.39/src/network.c:402: undefined reference to `bind'
/sources/lighttpd-1.4.39/src/network.c:418: undefined reference to `listen'
/sources/lighttpd-1.4.39/src/network.c:249: undefined reference to `socket'
/sources/lighttpd-1.4.39/src/network.c:234: undefined reference to `socket'
/sources/lighttpd-1.4.39/src/network.c:304: undefined reference to `getaddrinfo'
/sources/lighttpd-1.4.39/src/network.c:314: undefined reference to `freeaddrinfo'
/sources/lighttpd-1.4.39/src/network.c:327: undefined reference to `gethostbyname'
/sources/lighttpd-1.4.39/src/network.c:290: undefined reference to `setsockopt'
/sources/lighttpd-1.4.39/src/network.c:282: undefined reference to `in6addr_any'
/sources/lighttpd-1.4.39/src/network.c:305: undefined reference to `gai_strerror'
/sources/lighttpd-1.4.39/src/network.c:330: undefined reference to `__h_errno'
inet_ntop_cache.o: In function `inet_ntop_cache_get_ip':
/sources/lighttpd-1.4.39/src/inet_ntop_cache.c:31: undefined reference to `inet_ntop'
configfile-glue.o: In function `config_check_cond_nocache':
/sources/lighttpd-1.4.39/src/configfile-glue.c:383: undefined reference to `inet_aton'
http-header-glue.o: In function `http_response_redirect_to_directory':
/sources/lighttpd-1.4.39/src/http-header-glue.c:141: undefined reference to `getsockname'
/sources/lighttpd-1.4.39/src/http-header-glue.c:156: undefined reference to `getnameinfo'
/sources/lighttpd-1.4.39/src/http-header-glue.c:166: undefined reference to `inet_ntop'
/sources/lighttpd-1.4.39/src/http-header-glue.c:175: undefined reference to `gethostbyaddr'
/sources/lighttpd-1.4.39/src/http-header-glue.c:178: undefined reference to `__h_errno'
/sources/lighttpd-1.4.39/src/http-header-glue.c:180: undefined reference to `inet_ntoa'
collect2: error: ld returned 1 exit status
```
I guess we just need to tweak configure.ac and wanted to dig that but haven't done it yet.
BTW, sorry for the delay, I had hit the error but forgot to keep a copy of it. And then I switched to x86_gcc2. If you have any hints I'll be happy to help "unbreak" 1.4.39 on x86_64 and x86. Hmm, if making it cross-buildable is an option, I can try it.

pulkomandy replied:
Add -lnetwork to the LDFLAGS. Ideally the configure script should check for it using AC_SEARCH_LIBS.

fbrosson replied:
Great! It builds fine with a dirty workaround in the recipe (`LIBS=-lnetwork runConfigure ./configure --with-openssl`) but I'm trying to find a better fix that would only change configure.ac. Any hints (or even full directions)  are welcome :wink: Thanks!

pulkomandy commented on www-servers/lighttpd/lighttpd-1.4.39.recipe:
```diff
@@ -124,8 +125,9 @@ BUILD_PREREQUIRES="
+	autoconf
 	automake
-	runConfigure ./configure --with-openssl
+	LIBS=-lnetwork runConfigure ./configure --with-openssl
```
This can be done from inside configure.ac
https://autotools.io/autoconf/finding.html

something like this:

    AC_SEARCH_LIBS([gethostbyname], [network], [], [AC_MSG_ERROR([unable to find the gethostbyname() function])])

fbrosson commented:
I dont know if the recipe of 1.4.35 worked on x86 and x86_64, but what I had noticed was that the updated recipe for 1.4.39 did not work, _initialy_, on x86 and x86_64, although it worked for x86_gcc2 without any change other than the version bump.
Later, with your feedback, I was able to fix the recipe to make it work on most architectures (x86_gcc2, x86 and x86_64). So what we currently have in this PR is better than what we had for 1.4.35. But I think it still needs some changes because:
* all **`mod_*.so`** modules are installed in **`lib/`** instead of **`lib/lighttpd/`**
* **`BUILD()`** does not install any configuration files nor any default document (e.g. **`index.lighttpd.html`**)

I'll try to figure out what changes need to be made to fix those 2 issues. In the meanwhile, here are a few qustions:
1. Do you agree with my proposal to move all **`mod_*.so`** files into a lighttpd subdirectory ?
2. Would **`settings/network/lighttpd/`** be a good place to install the configuration? Or would it be better to use **`settings/extra/lighttpd/`** (or anything else more appropriate) ?
3. What would be the most appropriate directory to use for the default **`DOCUMENT_ROOT`** and should a default **`index.lighttpd.html`** be installed as non ediitable ?
4. Shall I use autoreconf instead of autoconf ?
5. Should I switch the patch suffix to «**`.patchset`**» (since I have undestood what happens behind the scenes and what needs to appear in the header of a patchset file) ?

Thanks!

pulkomandy commented:
1) sounds fine
2) settings/lighttpd. We don't usuqlly cqtegorize things further, currently (it works similar to /etc in Linux)
3) $dataDir/www , with a default read only document?
4) If autoconf works on its own, no need for autoreconf.
5) The process seems unclear to several people. Here is how it works:

- haikuporter -b to extract the sources and apply existing patches
- make changes as required inside work.../sources/
- run haikuporter again, iterate until the recipe builds
- the sources folder is a git repo, so commit your changes there (you can have multiple commits if you make several changes)
- use haikuporter --extract-patchset to generate the patchset file. No need to manually generate the patchset.

pulkomandy commented:
2) Yes, $settingsDir/lighttpd/ should be a directory with all lighttpd settings. What I meant is it should not be inside $settingsDir/network/lighttpd or similar.
3) maybe important to remember, $dataDir is not user writable so users will have to change the settings and set the correct directory. But I think it is a good thing to work that way. For cgi-bin, I don't know, technically they are executable and not libraries so I would use $binDir/cgi/ or so? But I'm fine with following Linux's packaging tradition as well.
4) For me it is a case by case thing. "autoreconf -fi" is a kind of bruteforce solution (regenerate everything). It's fine when it works and makes sure there are not osolete files leftover. The preferred solution is tu run the "autogen.sh" or "bootstrap.sh" sometimes provided in source arhives. But when there isn't one, we can go with "whatever works": just using pre-generated configure scripts, using autoreconf, manually re-runing some or all of the autotools, etc. It depends which versions of the tools we have and which are expected by the project.

fbrosson commented:
2) ok
3) I also like the idea of using a subdirectory of $dataDir. And I also agree that the fact that it is not user-writable makes it even more interesting. This increases security and makes the user responsible for chosing an alternate document_root.
4) ok, thanks!
Ah, I saw there are nice things I can take from the recipes of apache and nginx to improve that of lighttpd.

waddlesplash commented:
@fbrosson is this ready for review?

fbrosson commented:
Although the recipe already builds and works out of the box, I'd rather fix the most serious items in the TODO list before the review. I've been practicing with a few other recipes (in order to gain more experience in packaging for HaikuPorter) but I'll switch back to this one asap. The TODO items I plan to fix soon are the ones about:
a) replacing the "sed-created" index.shtml by a static version;
b) using manDir instead of $documentationDir/packages/man;
c) creating a self-signed certificate and enabling https in the default config.
I'll fix the other items during the review because I don't know yet how to fix them. But if anyone want's to test before I fix the most serious TODO items then, of course, there is no problem.

waddlesplash commented on www-servers/lighttpd/lighttpd-1.4.39.recipe:
```diff
@@ -110,41 +75,155 @@ BUILD_REQUIRES="
 TEST()
 {
 	make check
+	lighttpd -t -f /system/settings/lighttpd/lighttpd.conf
```
Delete this?

fbrosson replied:
Done.
Not really usefull indeed, as the resulting config file can only change if the recipe changes.
The only advantage would be to check that the runtime has been correctly built, but since most recipes don't have a real **`TEST()`** (other than a **`make check`**) adding an extra command here is not necessary.

waddlesplash commented:
> a) replacing the "sed-created" index.shtml by a static version;

Looked at recipe; this is OK as-is (and I might prefer it this way cuz we won't have to update the static version).

> b) using manDir instead of $documentationDir/packages/man;

Yeah, please do this.

> c) creating a self-signed certificate and enabling https in the default config.

Don't do this, let people do this themselves.

fbrosson commented:
Thanks for your feedback. I've updated the commit accordingly.
I've also made big changes to the devel package (instead of just dropping it) and updated the commit message. It now only contains all the .h files from the src subdir of the build directory, after the compilation. Is the devel package correct like that? Should I fix it (e.g. by cherry-picking only some header files) or simply drop it as I was initially planning to do?
Regarding the commit message, should I clean or modify it? Thanks!

fbrosson commented:
I've identified which header files (out off all lighttpd-14.39/src/*.h) can safely be removed from the devel package, and added a few commands in INSTALL() to do so. These files are:
* configparser.h
* network.h
* proc_open.h
* request.h

Do you think it is a good idea to not include them? I did not try (yet) to build a third-party lighttpd module with haikuporter (and the resulting devel package), but I checked the sources extensively and it seems that this list does not include any file that could be needed. I'll try, however, to test this.
If you prefer to include all headers or, alternatively, completely drop the devel package then I can fix the recipe for that.

fbrosson commented:
I've juste made some minor changes (added --with-zlib and --with-webdav-props with their dependencies).
BTW, I had forgotten to bump the revision. That's fixed now.

waddlesplash commented:
LGTM; if there's a problem with it we can always patch it later.
